In [37]:
import pandas as pd
import numpy as np

In [38]:
nl = pd.read_csv(r'C:\Users\David\Python\scrapy\facup\01_league_tables\nonleague_1990-11.csv')

In [39]:
nl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5054 entries, 0 to 5053
Data columns (total 42 columns):
A                              1234 non-null float64
A.1                            574 non-null float64
A.2                            574 non-null float64
AD                             24 non-null float64
AGA                            24 non-null float64
AGF                            24 non-null float64
AL                             24 non-null float64
AW                             24 non-null float64
Club                           110 non-null object
D                              5030 non-null float64
D.1                            574 non-null float64
F                              1234 non-null float64
F.1                            574 non-null float64
F.2                            574 non-null float64
GA                             3796 non-null float64
GD                             5054 non-null object
GF                             3796 non-null float64
HD        

In [40]:
nl.columns = nl.columns.str.lower()

In [41]:
unn_rows = ~nl['unnamed: 0'].isnull()  

nl[unn_rows]

a  a.1  a.2  ad  aga  agf  al  aw club   d  ...   unnamed: 0  \
0   40  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   6  ...            1   
1   58  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  10  ...            2   
2   39  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   7  ...            3   
3   50  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...            4   
4   79  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...            5   
5   42  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  12  ...            6   
6   48  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  14  ...            7   
7   66  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  12  ...            8   
8   55  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  14  ...            9   
9   58  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  15  ...           10   
10  64  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           11   
11  55  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           12   
12  61  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           13   
13  69  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  13  ...           14   
14  70  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...           15   
15  72  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           16   
16  61  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...           17   
17  82  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  10  ...           18   
18  77  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   5  ...           19   
19  82  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   7  ...           20   
20  86  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           21   
21  77  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  10  ...           22   
22  31  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...            1   
23  46  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   6  ...            2   
24  53  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  10  ...            3   
25  59  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  11  ...            4   
26  42  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...            5   
27  53  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  12  ...            6   
28  53  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  10  ...            7   
29  60  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  11  ...            8   
30  62  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...            9   
31  54  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  14  ...           10   
32  64  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  11  ...           11   
33  76  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   6  ...           12   
34  69  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...           13   
35  62  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  12  ...           14   
36  66  NaN  NaN NaN  NaN  NaN NaN NaN  NaN  11  ...           15   
37  64  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           16   
38  62  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   7  ...           17   
39  85  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   8  ...           18   
40  95  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   5  ...           19   
41  89  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   6  ...           20   
42  76  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   6  ...           21   
43  88  NaN  NaN NaN  NaN  NaN NaN NaN  NaN   9  ...           22   

               unnamed: 1  unnamed: 15  unnamed: 19  unnamed: 3 unnamed: 9  \
0          Colne Dynamoes          NaN          NaN         NaN        NaN   
1               Gateshead          NaN          NaN         NaN        NaN   
2           Witton Albion          NaN          NaN         NaN        NaN   
3             Hyde United          NaN          NaN         NaN        NaN   
4         South Liverpool          NaN          NaN         NaN        NaN   
5            Matlock Town          NaN          NaN         NaN        NaN   
6               Southport          NaN          NaN         NaN        NaN   
7          Fleetwood Town          NaN          NaN         NaN        NaN   
8                  Marine          NaN          NaN         NaN        NaN   
9             Bangor City          NaN          NaN         NaN        NaN   
10        Bishop Auckland          NaN   

For these league tables pos and team need copying into the correct column

In [42]:
nl.ix[unn_rows, 'team'] = nl.ix[unn_rows, 'unnamed: 1']
nl.ix[unn_rows, 'pos'] = nl.ix[unn_rows, 'unnamed: 0']

In [43]:
nl.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5054 entries, 0 to 5053
Data columns (total 42 columns):
a                              1234 non-null float64
a.1                            574 non-null float64
a.2                            574 non-null float64
ad                             24 non-null float64
aga                            24 non-null float64
agf                            24 non-null float64
al                             24 non-null float64
aw                             24 non-null float64
club                           110 non-null object
d                              5030 non-null float64
d.1                            574 non-null float64
f                              1234 non-null float64
f.1                            574 non-null float64
f.2                            574 non-null float64
ga                             3796 non-null float64
gd                             5054 non-null object
gf                             3796 non-null float64
hd        

Only need to keep year, league, post and team for this project, might clean other columns at a later stage

'team' seems to be missing 110 values, possibly in 'club'

In [44]:
nl[~nl.club.isnull()][['club', 'team']].head()

club team
1907          Tamworth  NaN
1908  Stafford Rangers  NaN
1909    Dover Athletic  NaN
1910     Tiverton Town  NaN
1911   Chippenham Town  NaN

In [45]:
nl[~nl.club.isnull()][['club', 'team']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 110 entries, 1907 to 2038
Data columns (total 2 columns):
club    110 non-null object
team    0 non-null object
dtypes: object(2)
memory usage: 2.6+ KB


The missing teams correspond with the club column

Filling missing teams into team column

In [46]:
nl['team'] = nl['club'].where(nl['team'].isnull(), other=nl['team'])

Removing unimportant columns

In [101]:
nl2 = nl[['team', 'pos', 'pts', 'year', 'league']].copy()

In [102]:
nl2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5054 entries, 0 to 5053
Data columns (total 5 columns):
team      5054 non-null object
pos       5054 non-null float64
pts       5054 non-null object
year      5054 non-null int64
league    5054 non-null object
dtypes: float64(1), int64(1), object(3)
memory usage: 236.9+ KB


In [103]:
nl2.pos.value_counts()

12    231
11    231
2     231
3     231
4     231
5     231
6     231
7     231
8     231
9     231
10    231
1     231
13    231
14    231
15    231
16    230
17    227
18    226
19    224
20    224
21    221
22    205
23     20
24     12
Name: pos, dtype: int64

In [104]:
nl2.year.value_counts()

2011    267
2010    266
2009    263
2008    256
2007    246
2006    237
2005    236
2003    221
2002    221
2001    221
2000    220
2004    220
1991    220
1990    219
1995    219
1992    219
1994    218
1999    218
1993    217
1998    217
1996    217
1997    216
Name: year, dtype: int64

In [105]:
nl2.team.value_counts()

Hendon !Hendon                                        22
Whyteleafe !Whyteleafe                                22
Harrow Borough !Harrow Borough                        22
Uxbridge !Uxbridge                                    22
Frickley Athletic !Frickley Athletic                  21
Billericay Town !Billericay Town                      20
Matlock Town !Matlock Town                            19
Horsham !Horsham                                      19
Barton Rovers !Barton Rovers                          19
Chorley !Chorley                                      19
Dulwich Hamlet !Dulwich Hamlet                        19
Tooting & Mitcham United !Tooting & Mitcham United    19
Metropolitan Police !Metropolitan Police              19
Ware !Ware                                            19
Heybridge Swifts !Heybridge Swifts                    19
Marine !Marine                                        19
Marlow !Marlow                                        19
Staines Town !Staines Town     

Removingbrackets (+ contents) from end of names, first checking there are no real team names that will be removed

In [106]:
nl2.team.str.split('(').str[1].unique()

array([nan, 'C)\xc2\xa0', 'P)', 'R)', 'C)', 'IOW)', 'O)\xc2\xa0', 'C) ',
       'O)', 'IOW) !Newport '], dtype=object)

In [107]:
nl2.team.str.split('[').str[1].unique()

array([nan, 'notes 1]', 'notes 4]', 'notes 5]'], dtype=object)

In [108]:
iow = nl2.team.str.contains('\(IOW\)')

nl2[iow]

team  pos pts  year  \
1690                 Newport (IOW)    4  76  1999   
4642  Newport (IOW) !Newport (IOW)   18  40  2005   
4706  Newport (IOW) !Newport (IOW)   22  29  2006   

                                          league  
1690  SOUTHERN FOOTBALL LEAGUE SOUTHERN DIVISION  
4642                ISTHMIAN LEAGUE DIVISION ONE  
4706                ISTHMIAN LEAGUE DIVISION ONE

Removing brackets and content but avoiding Newport (IOW)

In [116]:
nl2.loc[~iow, 'team'] = nl2.loc[~iow, 'team'].str.replace('\(.*', '') 
nl2.loc[~iow, 'team'] = nl2.loc[~iow, 'team'].str.replace('\[.*', '') 

In [117]:
nl2.team.str.split('(').str[1].unique()

array([nan, 'IOW)', 'IOW) !Newport '], dtype=object)

In [119]:
nl2.team.str.split('[').str[1].unique()

array([ nan])

Removing repeated team names e.g. Arsenal ! Arsenal is replaced with Arsenal

In [120]:
nl2.team = nl2.team.str.split('!').str[0]

In [121]:
nl2.team.value_counts()

Marlow                       22
Dulwich Hamlet               22
Chesham United               22
Billericay Town              22
Tooting & Mitcham United     22
Boreham Wood                 22
Walton & Hersham             22
Horsham                      22
Barton Rovers                22
Ware                         22
Uxbridge                     22
Leatherhead                  22
Heybridge Swifts             22
Windsor & Eton               22
Staines Town                 22
Harrow Borough               22
Whyteleafe                   22
Worthing                     22
Metropolitan Police          22
Hendon                       22
Hitchin Town                 22
Aveley                       22
Bromley                      21
Tilbury                      21
Frickley Athletic            21
Marine                       21
Chorley                      21
Bracknell Town               21
Ashford Town                 21
Matlock Town                 21
                             ..
Grimsby 

Regularising team names (stripping blank space and converting to UPPER case)

In [122]:
nl2['team'].replace('\xc2\xa0', '', inplace = True, regex = True) #removing 'non-breaking' spaces
nl2['team'] = nl2['team'].str.upper().str.strip() #converting to UPPER and stripping spaces

In [123]:
nl2.team.value_counts()

ASHFORD TOWN                  32
MATLOCK TOWN                  22
WEYMOUTH                      22
WORCESTER CITY                22
ALTRINCHAM                    22
HARROGATE TOWN                22
STAFFORD RANGERS              22
CHORLEY                       22
SLOUGH TOWN                   22
LEWES                         22
SUTTON COLDFIELD TOWN         22
ST ALBANS CITY                22
GRAYS ATHLETIC                22
AVELEY                        22
TOOTING & MITCHAM UNITED      22
LEEK TOWN                     22
MARLOW                        22
RADCLIFFE BOROUGH             22
MARINE                        22
CARSHALTON ATHLETIC           22
DULWICH HAMLET                22
CRAWLEY TOWN                  22
BISHOP'S STORTFORD            22
BASINGSTOKE TOWN              22
WELLING UNITED                22
BOREHAM WOOD                  22
BATH CITY                     22
BASHLEY                       22
REDDITCH UNITED               22
WORTHING                      22
          

In [124]:
nl2.to_csv(r'C:\Users\David\Python\scrapy\facup\01_league_tables\nonleague_1990-11_v2.csv', index=False)